### ISSUE TITLE
Managed to Unmanaged Performance Issues

### ISSUE THREAD

**Issue by optionj85** [CoraleStudios/Colore#31]
I have run into performance issues with CPU utilization.  I ran into this problem when trying to make use of the Custom Grid Effects.

I have found that NativeMethods.CreateKeyboardCustomGridEffects() appears to be the bottle-neck... the stack size is an unnecessary 300+ bytes.  A pointer should be passed instead to reduce this overhead. 

I believe the issue is that we are passing a managed struct from managed code to the unmanaged Razer SDK.  In theory, the .NET framework should probably do this better but not much help right now.

if we could marshal the CustomGrid struct to a pointer to unmanaged memory to an unmanaged method... we could then pass that structure to the Razer SDK in totally unmanaged code.

Perhaps an unmanaged wrapper for the Razer SDK could be placed between Colore and the Razer SDK to assist with correctly passing the CustomGrid as a pointer?  I guess the Razer SDK doesn't provide methods that accept pointers because they're expecting software to be written in C++?

Have I been clear enough? 

Any thoughts from anyone else?

Kind Regards
John Love
JL Software Development

**Comment by Sharparam** [comment]
The problem is Razer's SDK takes a struct, and not a pointer in their create function (as you guessed):

`RZRESULT CreateKeyboardCustomGridEffects(CUSTOM_GRID_EFFECT_TYPE Effects, RZEFFECTID * pEffectId)`

So we can't just pass in a pointer directly (that I'm aware of). So we'd either need to optimize the existing struct somehow, or have a bridge DLL that takes a pointer and deferences it before passing on to the SDK (but I dunno if that would even improve performance, it's worth testing though).

As I don't have a keyboard to test with, I can't directly test any changes and benchmark them. What is the speed/memory difference between custom grid operations in Colore with C# and a C++ app using the Chroma SDK directly?

And since I'm not very knowledgeable about the whole stack, heap et.c stuff, what exactly is the unnecessary data on the stack?

And are there similar performance issues with other functions that would warrant them being put in a bridge DLL as well? I would guess probably not, since the whole CustomGrid feature is what is the most complex to pass from managed to unmanaged, we even had to "hack" in a solution since multidimensional arrays can't be marshalled directly for some reason.

**Comment by optionj85** [comment]
Yes, the problem starts with the Razer SDK which only takes a struct as an argument.  I've already tried just modifying the import declaration to accept a pointer and whilst it works without error, the results are not as they should be.  Only some keys change color and they change to the wrong color.

Q: What is the speed/memory difference between custom grid operations in Colore with C# and a C++ app using the Chroma SDK directly?
A: I have not tested this but the level meter example that comes with the SDK seems to perform better and the fact that the rainbow scrolling effect is so smooth it would suggest that a much higher "framerate" is possible.

Q: And since I'm not very knowledgeable about the whole stack, heap et.c stuff, what exactly is the unnecessary data on the stack?
A: The unnecessary data on the stack is the color for each key.  The only data we should really be passing to the stack is a pointer to unmanaged memory that contains the structure with the CustomGrid in it.

Q: And are there similar performance issues with other functions that would warrant them being put in a bridge DLL as well?
A: I haven't tried the other functions since I didn't have the need to use them.  

My intention is to begin with a C++ library that has the same methods as the Razer SDK with the exception that the parameters are taken as pointers.  I'll only begin with the CustomGrid as this is what I initially need and can easily compare benchmarks.

I'll post the code up here when I have something working.

**Comment by Sharparam** [comment]
Sounds good. So wouldn't this mean that regular C++ apps are also passing the same unnecessary data since they are also passing structs by value (and thus copying the data) rather than a pointer? Or is C++ doing some magic here.

**Comment by optionj85** [comment]
I think it would mean that regular C++ apps are also passing the same unnecessary data, but I don't think the same overheads would occur.  I believe the initial overhead is due to the way the .NET framework handles calls between managed and unmanaged code.

This may be a bug with the .NET implementation but I can't be sure.

If my "bridge" DLL doesn't have the performance improvement that I require then I guess there are 2 options:
A) Converse with Razer to alert them to the issue of calling their API from an unmanaged assembly.  (Not sure how they would act on this or even how long it would take them)
B) Have a look at the disassembly of the SDK and see if we can safely implement the same features directly through a custom unmanaged library (removing the requirement for the Razer SDK entirely)

In short, let's experiment one step at a time and see where we get.

**Comment by Sharparam** [comment]
A is definitely possible, but I bet we'll have to wait a good while for a reply from them, and being a corporate business, who knows if they'll make the change.

Razer may need to be contacted for B as well, as we would effectively be reverse-engineering their SDK and implementing it ourselves, that feels like it'd be against their license or whatever. It may not play well alongside their own SDK or other apps either, and create headaches when they implement it in Synapse.

**Comment by optionj85** [comment]
I know I'm being slow on this but I'm really busy with work right now.

It's probably worth skipping A and trying this bridge DLL instead... just to see if we can eliminate the overheads for marshalling a .NET structure and passing it as a pointer to the bridge DLL instead.

With regards to actually replacing the Razer SDK, the IDA decompilation for the CreateKeyboardCustomGridEffects() method is as follows:

signed int __cdecl CreateKeyboardCustomGridEffects(CUSTOM_GRID_EFFECT_TYPE grid, void *id)
{
  signed int result; // eax@1
  CUSTOM_GRID_EFFECT_TYPE gridcopy; // [sp-214h] [bp-220h]@1

  memcpy(&gridcopy, &grid, sizeof(gridcopy));
  result = sub_100027F0(gridcopy, id);
  if ( result )
    result = 0x80004005u;
  return result;
}
[16/04/2015 14:52:24] Petemate: signed int __stdcall sub_100027F0(CUSTOM_GRID_EFFECT_TYPE grid, void *id)
{
  void *v2; // edi@1
  int v3; // esi@1
  int v5; // edx@7
  int v6; // eax@7
  int v7; // ecx@7
  GUID pguid; // [sp+10h] [bp-438h]@4
  char Dst; // [sp+20h] [bp-428h]@4
  char Src; // [sp+230h] [bp-218h]@1

  memcpy(&Src, &grid, 0x210u);
  v2 = id;
  v3 = dword_1000D650;
  if ( sub_10001930() && !(dword_1000D644 & (_DWORD )&Data) )
    return 4309;
  (_DWORD )&pguid.Data2 = dword_1000A270;
  pguid.Data1 = dword_1000A26C;
  (_DWORD )&pguid.Data4[0] = dword_1000A274;
  (_DWORD )&pguid.Data4[4] = dword_1000A278;
  memset(&Dst, 0, 0x210u);
  if ( CoCreateGuid(&pguid) >= 0 )
  {
    memcpy_s(&Dst, 0x210u, &Src, 0x210u);
    if ( sub_10004930((void *)(v3 + 1228), &Dst) )
    {
      if ( v2 )
      {
        v5 = (_DWORD )&pguid.Data2;
        v6 = (_DWORD )&pguid.Data4[0];
        (_DWORD )v2 = pguid.Data1;
        v7 = (_DWORD )&pguid.Data4[4];
        ((_DWORD )v2 + 1) = v5;
        ((_DWORD )v2 + 2) = v6;
        ((_DWORD )v2 + 3) = v7;
        return 0;
      }
      if ( !sub_10002DC0(pguid.Data1, (_DWORD )&pguid.Data2, (_DWORD )&pguid.Data4[0], (_DWORD )&pguid.Data4[4])
        && !sub_100034F0(pguid.Data1, (_DWORD )&pguid.Data2, (_DWORD )&pguid.Data4[0], (_DWORD )&pguid.Data4[4]) )
        return 0;
    }
  }
  return 1;
}

So after the data is copied by marshalling from managed to unmanaged code, the SDK then copied the data a second time on the following line:

memcpy(&gridcopy, &grid, sizeof(gridcopy));

With regards to the licensing I will have to check out the T's & C's but it wouldn't harm to develop something and present them with a suggestion if the licensing doesn't allow publication by a third-party.

Kind Regards
Option

**Comment by Sharparam** [comment]
(I added code formatting to your comment to make it more readable).

That decompilation info is interesting, makes things quite inefficient and I don't understand why they aren't just taking a pointer as a parameter directly...

**Comment by optionj85** [comment]
.Thanks, I'm kind of new to actually contributing on GitHub so I wasn't sure how to add the formatting.  In fact, I've edited my post and I think it's removed the formatting again.

I'm not sure what the rules are here but maybe there's a better way we can discuss this without spamming this forum?

**Comment by Sharparam** [comment]
You can add code by putting them inside the multi-line code tags, like this:

    ```
    code here
    ```

Rules are pretty much per-repository, it should be fine to have the discussion here. There's no other forum we can use anyway, really. (There's Razer insider but it doesn't have the niceties that GitHub does.)

You could join the [Gitter chat](https://gitter.im/CoraleStudios/Colore) for an IM-/IRC-like option if you want. It would suit better for quick exchange of smaller messages whereas this page might be better for longer, more detailed posts that bring important updates to the issue (like your information from decompiling).

**Comment by optionj85** [comment]
Thanks for the info Sharparam.

Do you have any experience with creating a library in the same format as the Razer SDK so it can be accessed in the same way that Colore accesses the Razer SDK?

My developer requirements haven't really had me stretch too far into C++.  I've developed some imaging methods in C++ in order to better manage the way byte transformations have happened but  my daily chores generally leave me working with C#, PHP, JavaScript and MSSQL.

**Comment by Sharparam** [comment]
Most C experience is from a single course at Uni. I've done some work in C++ but pretty minimal, nothing like a library. We were going to write [something similar for the Switchblade SDK](https://github.com/SharpBlade/SharpBlade.Emulator) (to enable software testing when hardware is not available) but we hit some road blocks about solving the issue about bi-directional communication between the bridge, WPF app, and the actual app and ceased work on it for the time being.

A bridge DLL for this project would probably be simpler to make, as it would just forward calls to the underlying SDK (with some work in between on some calls, like CustomGrids). An SDK replacement would be more work and would probably involve learning how USB interfaces work (if there isn't something else to interface with, SwitchBlade had a COM interface).

**Comment by brandonscott** [comment]
Any thoughts on what to do with this?  If we're not going to look at this within the next week, we should move this out to v4.0 or close it.

**Comment by Sharparam** [comment]
Move to v4.0, it more or less depends on @optionj85 as I don't currently possess the confidence required in C++ to dabble with performance improvements on a lower level.

**Comment by brandonscott** [comment]
Any thoughts on moving forward on this?  Is it a valid issue, or can we close this to stop moving milestone to milestone?

**Comment by Sharparam** [comment]
I'll just remove the milestone to leave it as an issue with no definite point to be implemented.

It's still relevant and having a bridge library might simplify some things on the C# side especially with the effects utilizing 2D arrays.

Problem is we need people with more experience in C++ and interop.

**Comment by Sharparam** [comment]
Closing due to lack of interest and ability in our current dev team to look into this properly.
